File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = {}
        res = []

        # loop through nums and fill hashmap
        for n in nums:
            counter[n] = counter.get(n, 0) + 1

        # need to use count as index and number as value
        newArray = [[] for _ in range(len(nums) + 1)]

        # fill array
        for i, v in counter.items():
            newArray[v].append(i)
        
        # loop through in reverse and fill res
        for i in range(len(newArray)-1, -1, -1):
            for j in newArray[i]:
                if len(res) == k:
                    return res
                res.append(j)
        
        return res
    

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        // need a result vector with same size as nums
        vector<int> res(nums.size());
        int prefix = 1;
        int postfix = 1;

        // first pass is prefix
        for (int i = 0; i < nums.size(); i++) {
            res[i] = prefix;
            prefix = prefix * nums[i];
        }
        // second pass is apply postfix (loop in reverse)
        for (int i = nums.size()-1; i > -1; i--) {
            res[i] = res[i] * postfix;
            postfix = postfix * nums[i];
        }

        return res;

    }
};
